### ISSUE TITLE
Alignment faults on AArch64

### ISSUE THREAD

**Issue by Ivan-Velickovic** [seL4/seL4#1339]
In the past when compiling code for AArch64 for an seL4 environment I have always passed `-mstrict-align` to the compiler to avoid alignment faults, this was because I understood that AArch64 does not allow unaligned access.

Yesterday, after getting an alignment fault with a Microkit-based program, we looked more into things and realised this was a misconception and that it is only in some cases that unaligned access leads to a fault.

My understanding now is that `SCTLR_EL1.M` has to be set and `SCTLR_EL1.A` has to be unset for unaligned access to be allowed. What's confusing me now though is that seL4 seems to do what I would expect yet we still get alignment faults unless we compile with `-mstrict-align` or equivalent.

Here you can see the bits being set for SCTLR:
https://github.com/seL4/seL4/blob/c5b23791ea9f65efc4312c161dd173b7238c5e80/src/arch/arm/64/head.S#L50-L54

and here you can see what `CR_ALIGN_SET` is:
https://github.com/seL4/seL4/blob/c5b23791ea9f65efc4312c161dd173b7238c5e80/src/arch/arm/64/head.S#L14-L24

`ALLOW_UNALIGNED_ACCESS` does not seem to be passed in the build-system anywhere and so by default it should be `1`.

Has anyone else run into alignment faults or can explain what's going on?

**Comment by Indanz** [comment]
Yes, don't do unaligned accesses on device memory. That also means, don't use memcpy or other stdlibc functions on device memory.

You want the device memory attribute, otherwise writes may be re-ordered and that is not what you want (because if that was okay, you should just use cached memory instead).

It would make sense for seL4 to expose some more device attributes, but not all hardware supports all combinations and Arm discouraged the more exotic combinations IIRC.

**Comment by Indanz** [comment]
Yes, the issue stays around and searchable when it's marked closed.

You're right it should be documented in the manual, it currently isn't. Re-opening the issue as a reminder for that. The confusing thing is that what seL4 calls device memory is nothing like the Arm meaning of device memory.

Are you mapping shared memory as uncached that you have so much problems with this? I think I've only ran into it when trying to use memset on uncached memory at init time.

**Comment by kent-mcleod** [comment]
> Are you mapping shared memory as uncached that you have so much problems with this? I think I've only ran into it when trying to use memset on uncached memory at init time.

The only case that's come up for me is when using dma pools. Uncached vs cached seems to be a tradeoff depending on the size of the transaction because the switch to kernel mode to perform a cache invalidate operation is more than the overhead of uncached memory mapping. EG the ethernet RX and TX descriptor rings can be mapped uncached while the packet buffers are mapped cached and use maintenance operations.


@midnightveil is right that its odd for the arm specific flags for memory mappings don't expose arm specific mapping attributes.  An RFC for adding this would be useful. We'd also like to be able to assign specific shareability attributes in multikernel-based systems and this would also require more control over mapping attributes.

**Comment by Indanz** [comment]
> The only case that's come up for me is when using dma pools. Uncached vs cached seems to be a tradeoff depending on the size of the transaction because the switch to kernel mode to perform a cache invalidate operation is more than the overhead of uncached memory mapping.

First, on aarch64 you can do all the needed cache operations in user space, no need for a system call.

Secondly, my measurements showed that using cached memory and doing an extra syscall would be faster than using uncached memory for TX packets sized 250 bytes. Uncached memory writes take on the order of 100 ns per write, which adds up very quickly.

> EG the ethernet RX and TX descriptor rings can be mapped uncached while the packet buffers are mapped cached and use maintenance operations.

Yes, I only ran into the unaligned issue when zeroing the descriptor rings. I had a stab at doing those cached as well, but ran out of time to get it working. But it was only one uncached memory write to the descriptor per TX, so not really worth spending much time on.

> [@midnightveil](https://github.com/midnightveil) is right that its odd for the arm specific flags for memory mappings don't expose arm specific mapping attributes. An RFC for adding this would be useful. We'd also like to be able to assign specific shareability attributes in multikernel-based systems and this would also require more control over mapping attributes.

I don't think anyone has objections against adding these, but is there any Arm hardware out there that could use outer shareable? Marking memory as unshared seems like a bad idea.

**Comment by midnightveil** [comment]
>  is right that its odd for the arm specific flags for memory mappings don't expose arm specific mapping attributes. An RFC for adding this would be useful. We'd also like to be able to assign specific shareability attributes in multikernel-based systems and this would also require more control over mapping attributes.

Another case where this is actually necessary for correctness: in discussing https://github.com/au-ts/sddf/issues/740, we realise that for VirtIO drivers the guest or hypervisor shares coherent ("NORMAL", cached) pages between guest&vmm for the virtio rings. @KurtWu10 brought up that shareability attributes come into to play here; whilst we can't control the not-seL4 side, if (seL4 as guest) the hypervisor expects inner-shareable, whilst seL4 is non-shareable (for non-smp) we have mismatched memory attributes and memory is no longer coherent, requiring cache operations on both sides; if (seL4 as host) the final shareability does depend on s1+s2 mappings combined, but the VMM would need to have mapped its version of the VirtIO queues with the same shareability the guests have. (see: ARM ARM DDI 0487 (version L.b), in B2.11 "Mismatched memory attributes" for the effects of this). It's possible this ends up being a non-issue in the current configuration due to how s1+s2 attributes are combined, but it's also possible it's not. (see #1631 which will change this).

Concrete example, based on [S2CombineS1Shareability](https://developer.arm.com/documentation/ddi0487/mc/9057-asl_shared_translation_attrs_s2combines1shareability).

```
VMM (stage 2-only, native thread): Non-shareable
Hypervisor stage 2 for VM: Non-shareable
Guest OS stage 1: Inner-shareable

Combined VM S1 + S2 => Inner-shareable. VMM => Non-shareable. That's mismatched shareability.
If seL4 manages the VMM, we have no control over those attributes.
```

Other direction:

```
VMM (stage 2-only, native thread): Inner-shareable
Hypervisor stage 2 for VM: Inner-shareable
Guest OS stage 1: Non-shareable

Combined VM S1 + S2 => Inner-shareable. VMM => Inner-shareable. In this case, everything is OK.
```

**Comment by Indanz** [comment]
Mismatched shareability is mostly a problem for SMP (and perhaps some IOMMU implementations). If the VMM and guest OS run on the same core, it is always coherent. Mismatched memory attributes does not automatically lead to incoherence, but incoherence can be one of the side effects. However, incoherence always occurs between different observers.

But I don't see any reason why seL4 shouldn't always map memory as inner-shareable and then this issue goes away. Then it's just inner- versus outer-shareable that user space might want to control. But there seems very little value in supporting outer-shareable cached memory, so it doesn't seem urgent. User space can either map it uncached and achieve the same result, or use the proper memory and cache synchronisation instructions to work-around any problems.
